app/command_autocomplete.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+// GetSuggestions returns suggestions for user input.
+func (a *App) GetSuggestions(commands []*model.AutocompleteData, userInput, roleID string) []model.Suggestion {
+	suggestions := []model.Suggestion{}
+	index := strings.Index(userInput, " ")
+	if index == -1 { // no space in user input
+		for _, command := range commands {
+			if strings.HasPrefix(command.Trigger, userInput) && (command.RoleID == roleID || roleID == model.SYSTEM_ADMIN_ROLE_ID || roleID == "") {
+				suggestions = append(suggestions, model.Suggestion{Hint: command.Trigger, Description: command.HelpText})
+			}
+		}
+		return suggestions
+	}
+	for _, command := range commands {
+		input := userInput[index+1:]
+		if command.Trigger == userInput[:index] && (command.RoleID == roleID || roleID == model.SYSTEM_ADMIN_ROLE_ID || roleID == "") {
```

Comment:
I might have gotten the exact logic wrong, but maybe make it less nested, like
```suggestion
		if command.Trigger != userInput[:index] {
			continue
		}
		if roleID != "" && command.RoleID != roleID && roleID != model.SYSTEM_ADMIN_ROLE_ID {
			continue
		}
```